crates/turbopack-dev/src/ecmascript/evaluate/chunk.rs

```diff
@@ -0,0 +1,265 @@
+    source_map::{GenerateSourceMap, GenerateSourceMapVc, OptionSourceMapVc},
+};
+use turbopack_ecmascript::{
+    chunk::{EcmascriptChunkPlaceable, EcmascriptChunkPlaceableVc},
+    utils::StringifyJs,
+    SourceMapAssetReferenceVc,
+};
+
+use crate::{ecmascript::list::reference::ChunkListReferenceVc, DevChunkingContextVc};
+
+/// An Ecmascript chunk that:
+/// * Contains the Turbopack dev runtime code; and
+/// * Evaluates a list of runtime entries.
+#[turbo_tasks::value(shared)]
+pub(crate) struct EcmascriptDevEvaluateChunk {
+    pub(super) chunking_context: DevChunkingContextVc,
+    pub(super) entry_chunk: ChunkVc,
+    pub(super) other_chunks: AssetsVc,
+    pub(super) evaluated_entries: EvaluatedEntriesVc,
+}
+
+#[turbo_tasks::value_impl]
+impl EcmascriptDevEvaluateChunkVc {
+    /// Creates a new [`EcmascriptDevEvaluateChunkVc`].
+    #[turbo_tasks::function]
+    pub fn new(
+        chunking_context: DevChunkingContextVc,
+        entry_chunk: ChunkVc,
+        other_chunks: AssetsVc,
+        evaluated_entries: EvaluatedEntriesVc,
+    ) -> Self {
+        EcmascriptDevEvaluateChunk {
+            chunking_context,
+            entry_chunk,
+            other_chunks,
+            evaluated_entries,
+        }
+        .cell()
+    }
+
+    #[turbo_tasks::function]
+    async fn code(self) -> Result<CodeVc> {
+        let this = self.await?;
+
+        let output_root = this.chunking_context.output_root().await?;
+        let chunk_path = self.ident().path().await?;
+        let chunk_public_path = if let Some(path) = output_root.get_path_to(&chunk_path) {
+            path
+        } else {
+            bail!(
+                "chunk path {} is not in output root {}",
+                chunk_path.to_string(),
+                output_root.to_string()
+            );
+        };
+
+        let other_chunks = this.other_chunks.await?;
+        let mut other_chunks_paths = Vec::with_capacity(other_chunks.len());
+        for other_chunk in &*other_chunks {
+            let other_chunk_path = &*other_chunk.ident().path().await?;
+            if let Some(other_chunk_path) = output_root.get_path_to(other_chunk_path) {
+                other_chunks_paths.push(other_chunk_path.to_string());
+            }
+        }
+
+        let runtime_module_ids = this
+            .evaluated_entries
+            .await?
+            .iter()
+            .map({
+                let chunking_context = this.chunking_context;
+                move |entry| async move {
+                    if let Some(placeable) = EcmascriptChunkPlaceableVc::resolve_from(entry).await?
+                    {
+                        Ok(Some(
+                            placeable
+                                .as_chunk_item(chunking_context.into())
+                                .id()
+                                .await?,
+                        ))
+                    } else {
+                        Ok(None)
+                    }
```

Comment:
The purpose of the `EvaluatableAsset` trait is mostly to act as a marker trait. It's up to the implementor to actually decide *where* it can be evaluated, by implementing specific traits such as `EcmascriptChunkPlaceable`. For instance, we could have Wasm evaluated entries, which we might want to handle differently.